From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 18 Mar 2016 17:20:30 +0100
Subject: x86/topology: Use total_cpus not nr_cpu_ids for logical packages
Git-commit: 3e8db2246b434c6b18a6a9f09904038bddcf76c7
Patch-mainline: v4.6-rc1
References: FATE#319705 (supporting patches needed for exit_box)
Signed-off-by: Tony Jones <tonyj@suse.de>

    x86/topology: Use total_cpus not nr_cpu_ids for logical packages
    
    nr_cpu_ids can be limited on the command line via nr_cpus=. That can break the
    logical package management because it results in a smaller number of packages,
    but the cpus to online are occupying the full package space as the hyper
    threads are enumerated after the physical cores typically.
    
    total_cpus is the real possible cpu space not limited by nr_cpus command line
    and gives us the proper number of packages.
    
    Reported-by: Mike Galbraith <umgwanakikbuti@gmail.com>
    Fixes: 1f12e32f4cd5 ("x86/topology: Create logical package id")
    Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
    Cc: Xiong Zhou <jencce.kernel@gmail.com>
    Cc: Ingo Molnar <mingo@kernel.org>
    Cc: Borislav Petkov <bp@alien8.de>
    Cc: Andreas Herrmann <aherrmann@suse.com>
    Link: http://lkml.kernel.org/r/alpine.DEB.2.11.1603181254330.3978@nanos

---
 arch/x86/kernel/smpboot.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/arch/x86/kernel/smpboot.c
+++ b/arch/x86/kernel/smpboot.c
@@ -325,9 +325,14 @@ static void __init smp_init_package_map(
 	 * By not including this we'll sometimes over-estimate the number of
 	 * logical packages by the amount of !present siblings, but this is
 	 * still better than MAX_LOCAL_APIC.
+	 *
+	 * We use total_cpus not nr_cpu_ids because nr_cpu_ids can be limited
+	 * on the command line leading to a similar issue as the HT disable
+	 * problem because the hyperthreads are usually enumerated after the
+	 * primary cores.
 	 */
 	ncpus = boot_cpu_data.x86_max_cores;
-	__max_logical_packages = DIV_ROUND_UP(nr_cpu_ids, ncpus);
+	__max_logical_packages = DIV_ROUND_UP(total_cpus, ncpus);
 
 	/*
 	 * Possibly larger than what we need as the number of apic ids per
